### PR TITLE
chore: Change new canister snapshot fields to be optional

### PR DESCRIPTION
**Pull request by michael-weigelt**
This PR improves https://github.com/dfinity/ic/pull/4596 by making the fields future proof. 

Since the previous PR has not been rolled out, it's fine to accept the BUF_BREAKING.

### CODE REVIEW

File: rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto [hunk 1]
```diff
@@ -22,7 +28,7 @@ message CanisterSnapshotBits {
   uint64 wasm_memory_size = 9;
   uint64 total_size = 10;
   repeated canister_state_bits.v1.Global exported_globals = 11;
-  optional uint64 global_timer_nanos = 12;
-  canister_state_bits.v1.OnLowWasmMemoryHookStatus on_low_wasm_memory_hook_status = 13;
+  optional CanisterTimer global_timer = 12;
```
Comment (schneiderstefan) [inline rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto:31]: I don't think you need this optional, as fields with a message type are optional by default.

File: rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs [hunk 2]
```diff
@@ -2115,7 +2115,7 @@ fn read_canister_snapshot_metadata_succeeds() {
     assert_eq!(metadata.wasm_module_size, uni_canister_wasm.len() as u64);
     assert_eq!(metadata.wasm_chunk_store.len(), 1);
     assert_eq!(metadata.certified_data, cert_data);
-    assert_eq!(metadata.global_timer, GlobalTimer::Active(timestamp));
+    assert_eq!(metadata.global_timer, Some(GlobalTimer::Active(timestamp)));
```
Comment (mraszyk) [inline rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs:2118]: could you please also assert on the low wasm memory hook?

File: rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto (hunk 1 above)
Comment (mraszyk) [reply rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto:31]: Here we also want to be able to write null, not just be able to read values with thing field omitted.

File: rs/state_layout/src/state_layout.rs [hunk 3]
```diff
@@ -2872,14 +2872,13 @@ impl TryFrom<pb_canister_snapshot_bits::CanisterSnapshotBits> for CanisterSnapsh
...
-            .unwrap_or_default();
-        let on_low_wasm_memory_hook_status =
-            OnLowWasmMemoryHookStatus::try_from(on_low_wasm_memory_hook_status).unwrap_or_default();
+        let global_timer = item.global_timer.map(CanisterTimer::from);
+        let on_low_wasm_memory_hook_status = item
+            .on_low_wasm_memory_hook_status
+            .map(pb_canister_state_bits::OnLowWasmMemoryHookStatus::try_from)
+            .map(Result::unwrap_or_default)
+            .map(OnLowWasmMemoryHookStatus::try_from)
+            .and_then(Result::ok);
```
Comment (dsarlis) [inline rs/state_layout/src/state_layout.rs:2881]: Tbh, I found the previous version a bit more readable.

First, we decode the protobuf wire value into a protobuf struct
```
let on_low_wasm_memory_hook_status_pb =
    pb_canister_state_bits::OnLowWasmMemoryHookStatus::try_from(
        item.on_low_wasm_memory_hook_status,
    )
    .unwrap_or_default()
```
then we convert from the protobuf struct into the rust struct

```
let on_low_wasm_memory_hook_status = 
    OnLowWasmMemoryHookStatus::try_from(on_low_wasm_memory_hook_status_pb).ok();
```

Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs:2881]: The `unwrap_or_default` in the first step means that if we read an invalid i32 from disk, we default instead of returning an error which seems counter-intuitive to me.

File: rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto (hunk 1 above)
Comment (mraszyk) [reply rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto:31]: > as fields with a message type are optional by default

doesn't this statement apply to enumerations instead?

File: rs/state_layout/src/state_layout.rs (hunk 3 above)
Comment (michael-weigelt) [reply rs/state_layout/src/state_layout.rs:2881]: > The unwrap_or_default in the first step means that if we read an invalid i32 from disk, we default instead of returning an error which seems counter-intuitive to me.

We default to variant 0 which means unused, and then fail in the try_from later.

Comment (dsarlis) [reply rs/state_layout/src/state_layout.rs:2881]: Well, that would never happen now that the field becomes `optional` as protobuf would decode a non-existing value during first upgrade to `None`. This would only be the case in the previous version of the code.

Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs:2881]: The return value of `pb_canister_state_bits::OnLowWasmMemoryHookStatus::try_from` is
```
std::result::Result<ic_protobuf::state::canister_state_bits::v1::OnLowWasmMemoryHookStatus, prost::UnknownEnumValue>
```
with the error case produced when observing, e.g., 42 as the i32 on disk, right?

Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs:2881]: i.e., corresponding to a corrupted on-disk serialization that should result in an error instead of defaulting

File: rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto (hunk 1 above)
Comment (dsarlis) [reply rs/protobuf/def/state/canister_snapshot_bits/v1/canister_snapshot_bits.proto:31]: Stefan is referring to https://protobuf.dev/programming-guides/proto3/#field-labels. I think technically we can omit the optional keyword now that the canister timer becomes a `message`. For protobuf it will still behave as a field that might or might not be set. So it could still cover our use case but as the documentation suggests, explicitly adding optional is recommended.

File: rs/state_layout/src/state_layout.rs (hunk 3 above)
Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs:2881]: > This would only be the case in the previous version of the code.

Even in the previous version of the code, I'd expect `pb_canister_state_bits::OnLowWasmMemoryHookStatus::try_from` to succeed and return Ok(Unspecified) in case of missing fields in the on-disk value and only return the error Err(...) if the field is present on disk with an invalid value, e.g., 42.

Comment (michael-weigelt) [reply rs/state_layout/src/state_layout.rs:2881]: >that would never happen now that the field becomes optional as protobuf would decode a non-existing value during first upgrade to None. 

Where is this conversion that turns an unknown `i32` into a `None`? If this is how the proto types are supposed to work, then whoever added `OnLowWasmMemoryHookStatus` forgot this conversion. Because what I am doing (using existing conversion logic) is manually converting from `Optional<i32>` to `proto::OnLowWasmMemoryHookStatus`, so I have to check variant 0 in the conversion from `proto::OnLowWasmMemoryHookStatus` to `OnLowWasmMemoryHookStatus`.

Comment (dsarlis) [reply rs/state_layout/src/state_layout.rs:2881]: I think we might want to use instead `try_from_option_field` to make things a bit more clear.

Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs:2881]: In particular, we don't want to get an error if the field is `None`:
```
/// Converts an optional proto field into a Rust type using `From` / `TryFrom`.
/// Returns `Err(ProtoMappingError::MissingField(field))` if the field is
/// `None`. 
pub fn try_from_option_field<F, T, E>(
    field: Option<F>,
    field_name: &'static str,
) -> Result<T, ProxyDecodeError>
```

**Comment by schneiderstefan** [comment]
In lieu of the improved upgrade/downgrade test, could you please run the test that failed earlier this week to make sure it's ok on upgrade.

**Review by schneiderstefan — APPROVED** [review]

**Comment by mraszyk** [comment]
> In lieu of the improved upgrade/downgrade test

I don't see any recent PR that would add canister snapshots to upgrade/downgrade tests: what am I missing?

**Comment by mraszyk** [comment]
I believe we need to take a canister snapshot [here](https://github.com/dfinity/ic/blob/722fb4ad0266ff94ff797b5f2f5a05d45de7b4f1/rs/tests/consensus/upgrade/upgrade_downgrade_app_subnet_test.rs#L95) of the canister deployed just before this line. We could use the same agent used to create the canister.

**Comment by mraszyk** [comment]
> That's why I am saying that it would be good to run the test that failed after the earlier PR proactively (//rs/ledger_suite/icrc1:icrc_ledger_suite_integration_golden_state_upgrade_downgrade_test_u256), as it imitates an upgrade test.

Are we sure that there will be a snapshot in the mainnet state (that could have changed since we ran the test last time) when we rerun the test now?

File: rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs (hunk 2 above)
Comment (michael-weigelt) [reply rs/execution_environment/src/execution_environment/tests/canister_snapshots.rs:2118]: done

File: rs/state_layout/src/state_layout.rs (hunk 3 above)
Comment (michael-weigelt) [reply rs/state_layout/src/state_layout.rs]: >This is my mental model: [...]

I agree except with the erroring out, is it worth doing that if there is an (impossible) bad value on disk? I'd rather view this as Unspecified/Unknown

**Comment by michael-weigelt** [comment]
>In lieu of the improved upgrade/downgrade test, could you please run the test that failed earlier this week to make sure it's ok on upgrade.

@schneiderstefan thanks for the reminder, it succeeds.

File: rs/state_layout/src/state_layout.rs (hunk 3 above)
Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs]: > is it worth doing that if there is an (impossible) bad value on disk

this is what happens shortly before the lines you touched:
```
        let binary_hash = match item.binary_hash {
            Some(hash) => {
                let hash: [u8; 32] =
                    hash.try_into()
                        .map_err(|e| ProxyDecodeError::ValueOutOfRange {
                            typ: "BinaryHash",
                            err: format!("Expected a 32-byte long module hash, got {:?}", e),
                        })?;
                Some(hash.into())
            }
            None => None,
        };
```

Comment (mraszyk) [reply rs/state_layout/src/state_layout.rs]: But this is supposed to be a corner case so I do believe we can clean it up later.

**Comment by schneiderstefan** [comment]
>Are we sure that there will be a snapshot in the mainnet state (that could have changed since we ran the test last time) when we rerun the test now?

No but it's better than nothing.

**Comment by mraszyk** [comment]
> > Are we sure that there will be a snapshot in the mainnet state (that could have changed since we ran the test last time) when we rerun the test now?
> 
> No but it's better than nothing.

Let's maybe also check that the test is still failing (with the current mainnet state) on the commit from before #4596 .

**Comment by michael-weigelt** [comment]
>Let's maybe also check that the test is still failing (with the current mainnet state) on the commit from before https://github.com/dfinity/ic/pull/4596 .

It does. So the signal we're getting is real
